homeassistant/components/smhi/config_flow.py

```diff
@@ -12,113 +12,82 @@
-                    )
-
-                self._errors[CONF_NAME] = "name_exists"
-            else:
-                self._errors["base"] = "wrong_location"
-
-        # If hass config has the location set and is a valid coordinate the
-        # default location is set as default values in the form
-        if (
-            not smhi_locations(self.hass)
-            and await self._homeassistant_location_exists()
-        ):
-            return await self._show_config_form(
-                name=HOME_LOCATION_NAME,
-                latitude=self.hass.config.latitude,
-                longitude=self.hass.config.longitude,
-            )
-
-        return await self._show_config_form()
-
-    async def _homeassistant_location_exists(self) -> bool:
-        """Return true if default location is set and is valid."""
-        # Return true if valid location
-        return (
-            self.hass.config.latitude != 0.0
-            and self.hass.config.longitude != 0.0
-            and await self._check_location(
-                self.hass.config.longitude, self.hass.config.latitude
-            )
-        )
-
-    def _name_in_configuration_exists(self, name: str) -> bool:
-        """Return True if name exists in configuration."""
-        return name in smhi_locations(self.hass)
+        errors: dict[str, str] = {}
 
-    async def _show_config_form(
-        self,
-        name: str | None = None,
-        latitude: float | None = None,
-        longitude: float | None = None,
-    ) -> FlowResult:
-        """Show the configuration form to edit location data."""
+        if user_input is not None:
+            lat = user_input[CONF_LATITUDE]
+            long = user_input[CONF_LONGITUDE]
+            if await async_check_location(self.hass, long, lat):
+                name = DEFAULT_NAME
+                if (
+                    lat == self.hass.config.latitude
+                    and long == self.hass.config.longitude
+                ):
+                    name = HOME_LOCATION_NAME
+
+                user_input[CONF_NAME] = name
+
+                await self.async_set_unique_id(f"{lat}-{long}")
+                self._abort_if_unique_id_configured()
+                return self.async_create_entry(title=name, data=user_input)
+
+            errors["base"] = "wrong_location"
+
+        for entry in self.hass.config_entries.async_entries(DOMAIN):
+            if (
+                entry.data[CONF_LATITUDE] == self.hass.config.latitude
+                and entry.data[CONF_LONGITUDE] == self.hass.config.longitude
+            ):
+                return self.async_show_form(
+                    step_id="user",
+                    data_schema=vol.Schema(
+                        {
+                            vol.Required(CONF_LATITUDE): cv.latitude,
+                            vol.Required(CONF_LONGITUDE): cv.longitude,
+                        }
+                    ),
+                    errors=errors,
+                )
```

Comment:
There is a bit of code duplication here.

Maybe make this another function, maybe `_async_has_entry_for_system_long_lat()` and then set the value to be passed to `default=` below instead

Reply:
Did do a slightly different approach than using a new function but removed the code duplication.